File: 01 - Defeat the Algorithm Interview/M200.LongestPalindromicSubstring.py
class Solution:
    """
    @param s: input string
    @return: the longest palindromic substring
    """
    def longestPalindrome(self, s):
        # write your code here
        result = ""
        if not s:
            return result

        for i in range(len(s)):
            sub = self.find_palindrome(s, i, i)
            if len(sub) > len(result):
                result = sub
            sub = self.find_palindrome(s, i, i + 1)
            if len(sub) > len(result):
                result = sub

        return result

    def find_palindrome(self, s, left, right):
        size = len(s)
        while left >= 0 and right < size and s[left] == s[right]:
            left -= 1
            right += 1

        return s[left + 1: right]

File: 01 - Defeat the Algorithm Interview/M200.LongestPalindromicSubstring.java
public class Solution {
    /**
     * @param s: input string
     * @return: the longest palindromic substring
     */
    public String longestPalindrome(String s) {
        // write your code here
        if (s == null || s.length() == 0) return "";

        int start = 0, len = 0, longest = 0;
        for (int i = 0; i < s.length(); i++) {
            len = findPalindromeLength(s, i, i);
            if (len > longest) {
                longest = len;
                start = i - len / 2;
            }
            len = findPalindromeLength(s, i, i + 1);
            if (len > longest) {
                longest = len;
                start = i - len / 2 + 1;
            }
        }

        return s.substring(start, start + longest);
    }

    private int findPalindromeLength(String s, int l, int r) {
        int len = 0;

        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            len += l == r ? 1 : 2;
            l--;
            r++;
        }

        return len;
    }
}
